rs_server/s3_storage_handler/s3_storage_handler.py

```diff
@@ -17,243 +24,719 @@
+        self.logger.handlers = []
+        self.logger.addHandler(logging.StreamHandler(sys.stdout))
+        self.logger.info("S3StorageHandler created !")
+
+        self.s3_client_mutex = Lock()
+        self.access_key_id = access_key_id
+        self.secret_access_key = secret_access_key
+        self.endpoint_url = endpoint_url
+        self.region_name = region_name
+        self.s3_client: boto3.client = None
+        if not self.connect_s3():
+            raise RuntimeError("The connection to s3 storage could not be made")
+
+    # get the s3 handler
+    def __get_s3_client(self, access_key_id, secret_access_key, endpoint_url, region_name):
+        """Retrieve or creates an S3 client instance.
+
+        Args:
+            access_key_id (str): The access key ID for S3 authentication.
+            secret_access_key (str): The secret access key for S3 authentication.
+            endpoint_url (str): The endpoint URL for the S3 service.
+            region_name (str): The region name.
+
+        Returns:
+            boto3.client: An S3 client instance.
+        """
+        # This mutex is needed in case of more threads accessing at the same time this function
+        with self.s3_client_mutex:
+            if self.s3_client:
+                return self.s3_client
+            client_config = botocore.config.Config(
+                max_pool_connections=100,
+                retries={"total_max_attempts": 10},
+            )
+            try:
+                s3_client = boto3.client(
+                    "s3",
+                    aws_access_key_id=access_key_id,
+                    aws_secret_access_key=secret_access_key,
+                    endpoint_url=endpoint_url,
+                    region_name=region_name,
+                    config=client_config,
+                )
+            except ClientError as e:
+                if e.response["Error"]["Code"] == "EntityAlreadyExists":
+                    self.logger.error("This clent already exists")
+                else:
+                    self.logger.error("Unexpected error")
+                return None
+
+        return s3_client
+
+    def connect_s3(self):
+        """Establishe a connection to the S3 service.
+
+        Returns:
+            bool: True if the connection is successful, False otherwise.
+        """
+        if self.s3_client is None:
+            self.s3_client = self.__get_s3_client(
+                self.access_key_id,
+                self.secret_access_key,
+                self.endpoint_url,
+                self.region_name,
+            )
+        if self.s3_client is None:
+            return False
+        return True
+
+    def disconnect_s3(self):
+        """Close the connection to the S3 service."""
+        if self.s3_client is not None:
+            self.s3_client.close()
+        self.s3_client = None
+
+    # delete a file from s3 by using the s3 handler
+    def delete_file_from_s3(self, bucket, s3_obj):
+        """Delete a file from S3.
```

Comment:
You should not have both a python comment and a docstring. You should remove the python comment and put it in the docstring e.g.
```
    def delete_file_from_s3(self, bucket, s3_obj):
        """delete a file from s3 by using the s3 handler.
```
This is done several times in your code.

Reply:
Solved